drivers/lora/sx126x.c

```diff
@@ -386,9 +407,15 @@ void SX126xIoRfSwitchInit(void)
 void SX126xReset(void)
 {
 	LOG_DBG("Resetting radio");
+#if defined(SX126X_GPIO)
 	gpio_pin_set(dev_data.reset, GPIO_RESET_PIN, 1);
 	k_sleep(K_MSEC(20));
 	gpio_pin_set(dev_data.reset, GPIO_RESET_PIN, 0);
+#elif defined(SX126X_STM32WL)
+	LL_RCC_RF_EnableReset();
+	k_sleep(K_MSEC(20));
+	LL_RCC_RF_DisableReset();
+#endif
```

Comment:
I don't like those switches scattered all over the driver. I would rather want a variant specific callbacks like:
```
struct sx126x_variant {
	int (*init)(struct sx126x_data *data);
	void (*wakeup)(struct sx126x_data *data);
	void (*reset)(struct sx126x_data *data);
	bool (*is_busy)(struct sx126x_data *data);
	int (*dio1_state)(struct sx126x_data *data);
	void (*irq_mask)(struct sx126x_data *data);
	void (*irq_unmask)(struct sx126x_data *data);
};
```
and then implement a variant with gpios:
```
static int sx126x_gpio_init(struct sx126x_data *data)
{
	gpio_init_callback(&data->dio1_irq_callback,
			   sx126x_dio1_irq_callback, BIT(GPIO_DIO1_PIN));
	if (gpio_add_callback(&data->dio1, &data->dio1_irq_callback) < 0) {
		LOG_ERR("Could not set GPIO callback for DIO1 interrupt.");
		return -EIO;
	}
	
	return 0;
};

struct sx126x_variant sx126x_gpio = {
    .init = sx126x_gpio_init,
    ...
};
```
and variant for stm32wl:
```
static int sx126x_stm32wl_init(struct sx126x_data *data)
{
	IRQ_CONNECT(DT_INST_IRQN(0),
		    DT_INST_IRQ(0, priority),
		    radio_isr, DEVICE_DT_INST_GET(0), 0);
	LL_EXTI_EnableIT_32_63(LL_EXTI_LINE_44);
	irq_enable(DT_INST_IRQN(0));
	
	return 0;
}

struct sx126x_variant sx126x_stm32wl = {
    .init = sx126x_stm32wl_init,
    ...
};
```
possibly each implementation should be in a different file (`sx126x_gpio.c` / `sx126x_standalone.c` and `sx126x_stm32wl.c`), conditionally compiled.

Reply:
So you are thinking about abstracting the variants. Could give it a stab if that's the consensus but doing the whole struct variant thing seems way overkill for what's left of the different code paths now that I got rid of the NSS bit, and since we only support one radio instance, it would bring overhead and no other advantages.

The current driver is already an abstraction, if we want to split the variant in a separate file I'd much rather do a `sx126x_common.c` for the common function, and then keep sx126x.c for the GPIO driver and do a separate one for the ST specific things.

Frankly I'm not sure that's worth the extra file, Kconfig options etc for the six SX126X_STM32WL checks that are left, but let me know what you think.

Reply:
It is much better now without NSS control, but I would still like to have "gpio vs stm32wl" differences grouped together. It is hard to follow functions if there is if-else on each step, including doing so in preprocessor. What if we would create such functions:
```
#if defined(SX126X_GPIO)

static int sx126x_variant_init(struct sx126x_data *data)
{
	gpio_init_callback(&data->dio1_irq_callback,
			   sx126x_dio1_irq_callback, BIT(GPIO_DIO1_PIN));
	if (gpio_add_callback(&data->dio1, &data->dio1_irq_callback) < 0) {
		LOG_ERR("Could not set GPIO callback for DIO1 interrupt.");
		return -EIO;
	}
	
	return 0;
};

... other functions for gpio variant

#elif defined(SX126X_STM32WL)

static int sx126x_variant_init(struct sx126x_data *data)
{
	IRQ_CONNECT(DT_INST_IRQN(0),
		    DT_INST_IRQ(0, priority),
		    radio_isr, DEVICE_DT_INST_GET(0), 0);
	LL_EXTI_EnableIT_32_63(LL_EXTI_LINE_44);
	irq_enable(DT_INST_IRQN(0));
	
	return 0;
}

... other functions for stm32wl variant

#endif
```

Then we would meet somewhere in the middle, having variant specific grouped together and have no overhead on the runtime and codesize (functions will most likely inline by compiler).

There is also little work to check how much full abstraction would cost us with my initial idea with `struct sx126x_variant`. Hard to say how much overhead it is without trying it out.

Reply:
Yeah that sounds good, I added a new patch to shuffle the code around and group few functions together, and now the WL ones are in a single block. On the work handler I kept the IRQ renable line for both variants, it should be a noop on the normal one, so we don't have to have an extra case there. At this point it should be easy to add a new Kconfig option and split it into three files, I'd rather keep them on the same file though since we only have two options... less likely for someone to change one and forget the other.